# Initial GPEs 

We compute the $R^2$ and $\% ISE$ values:

In [26]:
import postprocessing

R2score_vec, ISE_vec = postprocessing.compute_R2_ISE()


Statistics on test set for GPE trained for the output LVV:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output RVV:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output LAV:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output RAV:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output LVOTdiam:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output RVOTdiam:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output LVmass:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output LVWT:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output LVEDD:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output SeptumWT:
  R2 = 1.00
  %ISE = 100.00 %


Statistics on test set for GPE trained for the output RVlongdiam:
  

> The maximum scores were achieved for all the emulators, except for the emulator of the TAT that achieved \(R^2=0.99\) and TATLV\textsubscript{endo} that achieved a \(ISE=98.86\%\). 

# Global Sensitivity Analysis

> First-order effects of the GSA when modifying anatomical and functional parameters are presented in Fig 5.4.

<img src="./figures/GSA1_small.png"  width="70%" />

The image above was built in PPT, putting together individual pies. To print all the single pies, run:

In [27]:
import postprocessing
import numpy as np

input_labels=np.genfromtxt("/media/crg17/Seagate Expansion Drive/param_labels.txt", dtype="str", delimiter="\n")
postprocessing.full_GSA(subfolder="initial_sweep", output_labels_dir="/media/crg17/Seagate Expansion Drive/biomarkers_labels.txt", input_labels=input_labels, first_order=True, second_order=False)

The results are in ```/media/crg17/Seagate Expansion Drive/initial_sweep/figures```

> In Fig \ref{fig:GSA2}, we show the second-order interactions for each output.

<img src="./figures/GSA2_small.png"  width="70%" />

The image above was built in PPT, putting together individual networks. The text in the centre of each network is also written manually. To print all the single networs, run:

In [ ]:
import postprocessing
import numpy as np

input_labels=np.genfromtxt("/media/crg17/Seagate Expansion Drive/param_labels.txt", dtype="str", delimiter="\n")
postprocessing.full_GSA(subfolder="initial_sweep", output_labels_dir="/media/crg17/Seagate Expansion Drive/biomarkers_labels.txt", input_labels=input_labels, first_order=False, second_order=True)

The results are in ```/media/crg17/Seagate Expansion Drive/initial_sweep/figures```

# Calibration to literature data

> We extracted these biomarkers from the NORRE study\cite{Kou2014EchocardiographicStudy}, and ran three waves of BHM.

To run the three waves using literature data (* **heavy computationally** *):

In [ ]:
import pipeline_bhm

pipeline_bhm.literature(run_wave0=True, run_wave1=True, run_wave2=True)

>By the last wave, the original space was reduced to \(9.55\%\) of the original space size, with a maximum uncertainty quotient of \(1.9\), and a median value of \(0.67\).

To obtain the numbers above we run:

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/Seagate Expansion Drive/initial_sweep/NROY_rel_literature.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/Seagate Expansion Drive/literature/wave1/NROY_rel_literature.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/Seagate Expansion Drive/literature/wave2/NROY_rel_literature.dat", dtype=float)

total_reduction = round(wave0_NROY_rel*1e-4*wave1_NROY_rel*wave2_NROY_rel,2)

uncertainty_quotient_vec = np.genfromtxt("/media/crg17/Seagate Expansion Drive/literature/wave2/variance_quotient_wave2_literature.dat", dtype=float)

print(total_reduction)
print(np.max(uncertainty_quotient_vec))
print(np.median(uncertainty_quotient_vec))

> In Fig 5.6 we show how the biomarkers obtained when simulating and emulating in the NROY region fall better within the range reported in the literature, as we run more waves.

<img src="./figures/output_evolution_small.png"  width="70%" />

The image above was built in PPT, putting together individual plots. To print all the single plots, run:

In [ ]:
import postprocessing

postprocessing.plot_output_evolution_seaborn(first_wave = 0, last_wave = 2,
                                   subfolder = "literature", only_feasible = False,
                                   output_labels_dir = "/media/crg17/Seagate Expansion Drive/biomarkers_labels.txt",
                                   exp_mean_name = "anatomy_EP_lit_mean.txt",
                                   exp_std_name = "anatomy_EP_lit_sd.txt",
                                   units_dir = "/media/crg17/Seagate Expansion Drive/biomarkers_units.txt")

The results are in ```/media/crg17/Seagate Expansion Drive/literature/figures```

# Calibration to synthetic patient data

To run the four waves using data from subject #01 (* **heavy computationally** *):

In [ ]:
import pipeline_bhm

pipeline_bhm.patient(patient_number=1, run_wave0=True, run_wave1=True, run_wave2=True, run_wave3=True, sd_magnitude=10)

> Compared to the initial space, the biggest reduction in NROY size happened in the first wave, with a reduction down to \(11.06\%\) of the original space. In the third wave, the space was reduced to \(8.89\%\) of the original size.

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient1_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave1/NROY_rel_patient1_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave2/NROY_rel_patient1_sd_10.dat", dtype=float)

print(wave0_NROY_rel)
print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.),2))
print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2))

> Using as ground truth the biomarkers extracted from that mesh, the implausibility of that point in the third wave was \(0.9\).

In [ ]:
implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave2/implausibilities_patient1_sd_10.dat", dtype=float)

print(np.max(implausibility))

> In the case of subject \(\#01\) in the third wave, although the maximum value of the variance quotient across all space was \(2.23\), the median value was \(1.08\), suggesting that the uncertainties of the emulators are comparable to the uncertainties of the ground truth data.

In [ ]:
VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave2/variance_quotient_wave2_patient1_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

## Reusing emulators with a different subject

> Firstly we ran the BHM pipeline for \(3\) waves with patient \(\#02\).

To run the three waves using data from subject #02 (* **heavy computationally** *):

In [ ]:
import pipeline_bhm

pipeline_bhm.patient(patient_number=2, run_wave0=False, run_wave1=False, run_wave2=True, run_wave3=False, sd_magnitude=10)

> In this case the NROY size was \(9.49\%\) of the original space, and we achieve a final NROY size of \(7.51\%\) of the original space.

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient2_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave1/NROY_rel_patient2_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/NROY_rel_patient2_sd_10.dat", dtype=float)

print(wave0_NROY_rel)
print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2))

> The implausibility for the known target parameter point was of \(0.2\),

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/implausibilities_patient2_sd_10.dat", dtype=float)

print(np.max(implausibility))

> and the maximum and median variance quotient (as defined in Eq \ref{eq:VQ}) were of \(2.2\) and \(0.72\), respectively.

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/variance_quotient_wave2_patient2_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

> We then used the emulators from the third wave of subject \(\#01\), but using the ground truth data (the simulation results) from subject \(\#02\).

In [ ]:
import pipeline_bhm

pipeline_bhm.mix_patients(use_emulators_from_patient=1, new_patient=2, sd_magnitude=10)

>Using the previously calibrated emulator, we obtained a NROY size of \(7.13\%\) of the original size without running any extra waves.

In [ ]:
import numpy as np

NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/NROY_rel_patient2_sd_10_lhd_1000000.dat", dtype=float)

print(NROY_rel)

> The implausibility of the point was \(0.7\)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/implausibilities_patient2_sd_10.dat", dtype=float)

print(np.max(implausibility))

> The uncertainties were smaller, with a maximum and median of \(1.94\) and \(0.94\), respectively.

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/variance_quotient_wave2_patient2_using_patient1_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

> Conversely, we tested the emulators from the third wave of subject \(\#02\), but on subject \(\#01\).

In [ ]:
import pipeline_bhm

pipeline_bhm.mix_patients(use_emulators_from_patient=2, new_patient=1, sd_magnitude=10)

> Using these already trained emulators, without running any extra waves, we obtained a NROY size of \(8.34\%\) of the original size (only \(0.55\) points of difference compared to the full pipeline).

In [ ]:
import numpy as np

NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient2_sd_10/wave2/NROY_rel_patient1_sd_10_lhd_1000000.dat", dtype=float)

print(NROY_rel)

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient2_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave1/NROY_rel_patient2_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/NROY_rel_patient2_sd_10.dat", dtype=float)

NROY_rel_full_pipeline = round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2)

print(round(NROY_rel - NROY_rel_full_pipeline,2))

> The implausibility of the point was \(0.7\)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient2_sd_10/wave2/implausibilities_patient1_sd_10.dat", dtype=float)

print(np.max(implausibility))

> and the maximum and median uncertainties were of \(2.51\) and \(0.82\), respectively.

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient2_sd_10/wave2/variance_quotient_wave2_patient1_using_patient2_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

|  | Subject | Using results from #01 | Using results from #02
| :-: | 
| NROY size as % of original space | #01 | 8.89 | 8.34 |
| NROY size as % of original space | #02 | 7.13 | 7.51 |
| Implausibility of simulated point | #01 | 0.9 | 0.7 |
| Implausibility of simulated point | #02 | 0.7 | 0.2 |
| Uncertainty quotient Max/median | #01 | 2.23 / 1.08 | 2.51 / 0.82 |
| Uncertainty quotient Max/median | #02 | 1.94 / 0.94 | 2.2 / 0.72 |
| NROY+RO / NROY match | #01 | - | 98.95% / 95.75% |
| NROY+RO / NROY match | #02 | 99.22% / 96.29% | - |


The numbers above have been mostly generated in previous snippets. The missing numbers are the NROY+RO / NROY match. They are calculated in the following snippet:

In [ ]:
import postprocessing

perc_1_using_2_RO = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=1, original_last_wave=2, using_patient=2,
                        using_last_wave=2)
perc_1_using_2_NROY = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=1, original_last_wave=2, using_patient=2,
                        using_last_wave=2)
perc_2_using_1_RO = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=2, original_last_wave=2, using_patient=1,
                        using_last_wave=2)
perc_2_using_1_NROY = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=2, original_last_wave=2, using_patient=1,
                        using_last_wave=2)


print(round(perc_1_using_2_RO,2), round(perc_1_using_2_NROY,2), round(perc_2_using_1_RO,2), round(perc_2_using_1_NROY,2))

## Reusing emulators with very different subjects

> In the first scenario, we computed the $\ell1$ distance between the input vector of subject \(\#01\) and the input vector of each other subject of the cohort

In [ ]:
import CT_cohort

CT_cohort.generate_patient_distances()

> reusing the third wave of patient \(\#01\) using the biomarkers of this new subject, lead to similar results as running the full pipeline.

In [ ]:
import pipeline_bhm

pipeline_bhm.run_farthest_patients(patient_number=1,input_or_output="input")

> The NROY space was of \(11.5\%\) without running any extra waves in the reusing scenario,

In [ ]:
import numpy as np

NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/NROY_rel_patient10_sd_10_lhd_1000000.dat", dtype=float)

print(NROY_rel)

> compared to \(12.2\%\) in the full-pipeline scenario (after three waves)

In [ ]:
wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient10_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave1/NROY_rel_patient10_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave2/NROY_rel_patient10_sd_10.dat", dtype=float)

NROY_rel_full_pipeline = round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2)

print(NROY_rel_full_pipeline)

> the implausibility for that subject was of \(0.39\)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/implausibilities_patient10_sd_10.dat", dtype=float)

print(np.max(implausibility))

> and the uncertainty quotients had a maximum of \(2.44\) and a median \(1.1\)

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/variance_quotient_wave2_patient10_using_patient1_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

> compared to an implausibility of \(0.34\)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave2/implausibilities_patient10_sd_10.dat", dtype=float)

print(np.max(implausibility))

> and uncertainty quotients with a maximum of \(2.02\) and median of \(1.18\) if using the specialised pipeline.

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave2/variance_quotient_wave2_patient10_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

> In the second scenario, we computed the $\ell1$ distance between the vector of biomarkers of subject \(\#01\) and the vector of biomarkers of each other subject in the cohort. The farthest subject in this case was subject \(\#18\).

In [ ]:
import pipeline_bhm

pipeline_bhm.run_farthest_patients(patient_number=1,input_or_output="output")

> Using the results of the last wave of \(\#01\) with the biomarkers of \(\#18\) resulted in a reduction of the original space down to \(1.94\%\).

In [ ]:
import numpy as np

NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/NROY_rel_patient18_sd_10_lhd_1000000.dat", dtype=float)

print(NROY_rel)

> The uncertainty quotient in the third wave of \(\#18\) had a maximum value of \(3.64\) and a median value of \(1.64\).

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/variance_quotient_wave2_patient18_using_patient1_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

> The lowest match in terms of the NROY region similarity was also achieved when using the third wave of \(\#01\) with the biomarkers of \(\#18\): \(83.61\%\) of the NROY points in the case of using the emulators of \(\#01\) with the biomarkers of \(\#18\) had the same status (not implausible) as with the case of running three waves using only case \(\#18\).

In [ ]:
import postprocessing

perc = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=18, original_last_wave=2, using_patient=1,
                        using_last_wave=2)

print(round(perc,2))

|  | NROY size as % of original space | Implausibility of simulated point | Uncertainty quotient Max/median | NROY+RO / NROY match
| :-: | 
| Using #10 on #01 | 8.73 | 0.7  |  2.43/1.31 | 99.06%/97.84% |
| Ground truth | 8.89 |  0.9 | 2.23/1.08  | - |
| Using #18 on #01 | 10.82 | 0.43 | 3.23/1.32  | 97.15%/97.71% |
| Ground truth | 8.89 |  0.9 | 2.23/1.08  | - |
| Using #01 on #10 | 11.5 | 0.39  | 2.44/1.1  | 98.97%/95.92% |
| Ground truth | 12.2 | 0.34 |  2.02/1.18 | - |
| Using #01 on #18 |  1.94 | 0.93  | 3.64/1.64 | 99.33%/83.61% |
| Ground truth | 2.15 | 0.48 | 3.93/1.70  | - |

Most of the numbers in the table have been already computed in previous snippets. The numbers of the first row:

In [ ]:
import numpy as np

NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient10_sd_10/wave2/NROY_rel_patient1_sd_10_lhd_1000000.dat", dtype=float)

print(NROY_rel)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient10_sd_10/wave2/implausibilities_patient1_sd_10.dat", dtype=float)

print(np.max(implausibility))

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient10_sd_10/wave2/variance_quotient_wave2_patient1_using_patient10_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

In [ ]:
import postprocessing

perc1 = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=1, original_last_wave=2, using_patient=10,
                        using_last_wave=2)
perc2 = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=1, original_last_wave=2, using_patient=10,
                        using_last_wave=2)

print(round(perc1,2),round(perc2,2))

For the third row:

In [ ]:
import numpy as np

NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient18_sd_10/wave2/NROY_rel_patient1_sd_10_lhd_1000000.dat", dtype=float)

print(NROY_rel)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient18_sd_10/wave2/implausibilities_patient1_sd_10.dat", dtype=float)

print(np.max(implausibility))

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient18_sd_10/wave2/variance_quotient_wave2_patient1_using_patient18_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

In [ ]:
import postprocessing

perc1 = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=1, original_last_wave=2, using_patient=18,
                        using_last_wave=2)
perc2 = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=1, original_last_wave=2, using_patient=18,
                        using_last_wave=2)

print(round(perc1,2),round(perc2,2))

NROY match using #01 on #10:

In [ ]:
import postprocessing

perc1 = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=10, original_last_wave=2, using_patient=1,
                        using_last_wave=2)
perc2 = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=10, original_last_wave=2, using_patient=1,
                        using_last_wave=2)

print(round(perc1,2),round(perc2,2))

Implausibility of simulated point using #01 on #18:

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/using_patient1_sd_10/wave2/implausibilities_patient18_sd_10.dat", dtype=float)

print(np.max(implausibility))

In [ ]:
import postprocessing

perc = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=18, original_last_wave=2, using_patient=1,
                        using_last_wave=2)

print(round(perc,2))

Last row of the table:

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient18_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave1/NROY_rel_patient18_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave2/NROY_rel_patient18_sd_10.dat", dtype=float)

NROY_rel_full_pipeline = round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2)

print(NROY_rel_full_pipeline)

In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave2/implausibilities_patient18_sd_10.dat", dtype=float)

print(np.max(implausibility))

In [ ]:
import numpy as np

VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave2/variance_quotient_wave2_patient18_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

# Modifying the accuracy of the data

We see the effect now of using SD=5% of the mean. It should make the NROYs more personalised, and so increase the discrepancy between literature and personalised.

In [ ]:
import pipeline_bhm

pipeline_bhm.patient(patient_number=1, run_wave0=True, run_wave1=True, run_wave2=True, run_wave3=True, sd_magnitude=5)

In [ ]:
import pipeline_bhm

pipeline_bhm.patient(patient_number=18, run_wave0=True, run_wave1=True, run_wave2=True, run_wave3=True, sd_magnitude=5)

# Reducing the uncertainty of the simulations

> In the case of subject #01, the final NROY size went from 10.18% of the original space, in the last wave of the case of high uncertainty, down to 0.43%, in the first wave of the case of low uncertainty. By the third wave, the NROY space is 0.000096% of the original space. 

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient1_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave1/NROY_rel_patient1_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave2/NROY_rel_patient1_sd_10.dat", dtype=float)

print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.*wave3_NROY_rel/100.),2))

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep_postthesis/NROY_rel_patient1_sd_5_lhd_100000.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_5/wave1/NROY_rel_patient1_sd_5_lhd_100000.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_5/wave2/NROY_rel_patient1_sd_5_lhd_100000.dat", dtype=float)
print(float(wave0_NROY_rel))
print(float(wave0_NROY_rel*wave1_NROY_rel/100.))
print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.*wave3_NROY_rel/100.),6))

> Similarly, subject #02’s NROY changed from 7.26% to 0.00003%

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient2_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave1/NROY_rel_patient2_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/NROY_rel_patient2_sd_10.dat", dtype=float)

print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2))

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient2_sd_5_lhd_100000.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_5/wave1/NROY_rel_patient2_sd_5_lhd_100000.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_5/wave2/NROY_rel_patient2_sd_5_lhd_100000.dat", dtype=float)
print(float(wave0_NROY_rel))
print(float(wave0_NROY_rel*wave1_NROY_rel/100.))
print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100),6))

> subject #10, from 11.79% to 0.00003%

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep_thesis/NROY_rel_patient10_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave1/NROY_rel_patient10_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave2/NROY_rel_patient10_sd_10.dat", dtype=float)

print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.),2))

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient10_sd_5_lhd_100000.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_5/wave1/NROY_rel_patient10_sd_5_lhd_100000.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_5/wave2/NROY_rel_patient10_sd_5_lhd_100000.dat", dtype=float)
print(float(wave0_NROY_rel))
print(float(wave0_NROY_rel*wave1_NROY_rel/100.))
print(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.))

> and subject #18, from 2.07% to 0.13%. 

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient18_sd_10.dat", dtype=float)
wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave1/NROY_rel_patient18_sd_10.dat", dtype=float)
wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave2/NROY_rel_patient18_sd_10.dat", dtype=float)

print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.*wave3_NROY_rel/100.),2))

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep_postthesis/NROY_rel_patient18_sd_5_lhd_100000.dat", dtype=float)
#wave1_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_5/wave1/NROY_rel_patient18_sd_5_lhd_100000.dat", dtype=float)
#wave2_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_5/wave2/NROY_rel_patient18_sd_5_lhd_100000.dat", dtype=float)
print(float(wave0_NROY_rel))
#print(float(wave0_NROY_rel*wave1_NROY_rel/100.))
#print(round(float(wave0_NROY_rel*wave1_NROY_rel/100.*wave2_NROY_rel/100.*wave3_NROY_rel/100.),6))

> In terms of VQ, the biggest increase was observed in subjects #01 where the maximum value of VQ changed from 2.92 to 14.03 and the median value changed from 1.11 to 8. 

In [ ]:
VQ_p1_10 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave2/variance_quotient_wave2_patient1_sd_10.dat", dtype=float)

VQ_p1_5 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_5/wave2/variance_quotient_wave2_patient1_sd_5.dat", dtype=float)

VQ_p2_10 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/variance_quotient_wave2_patient2_sd_10.dat", dtype=float)

VQ_p2_5 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_5/wave2/variance_quotient_wave2_patient2_sd_5.dat", dtype=float)

VQ_p10_10 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave2/variance_quotient_wave2_patient10_sd_10.dat", dtype=float)

VQ_p10_5 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_5/wave2/variance_quotient_wave2_patient10_sd_5.dat", dtype=float)

VQ_p18_10 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave2/variance_quotient_wave2_patient18_sd_10.dat", dtype=float)

VQ_p18_5 = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/variance_quotient_wave0_patient18_sd_5.dat", dtype=float)

print(np.max(VQ_p1_5)-np.max(VQ_p1_10))
print(np.median(VQ_p1_5)-np.median(VQ_p1_10))

print(np.max(VQ_p2_5)-np.max(VQ_p2_10))
print(np.median(VQ_p2_5)-np.median(VQ_p2_10))

print(np.max(VQ_p10_5)-np.max(VQ_p10_10))
print(np.median(VQ_p10_5)-np.median(VQ_p10_10))

print(np.max(VQ_p18_5)-np.max(VQ_p18_10))
print(np.median(VQ_p18_5)-np.median(VQ_p18_10))

print(np.max(VQ_p1_10))
print(np.max(VQ_p1_5))
print(np.median(VQ_p1_10))
print(np.median(VQ_p1_5))
print(np.max(VQ_p2_10))
print(np.max(VQ_p2_5))
print(np.median(VQ_p2_10))
print(np.median(VQ_p2_5))


In [ ]:
import numpy as np

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_10/wave2/implausibilities_patient1_sd_10.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient1_sd_5/wave2/implausibilities_patient1_sd_5.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_10/wave2/implausibilities_patient2_sd_10.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient2_sd_5/wave2/implausibilities_patient2_sd_5.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_10/wave2/implausibilities_patient10_sd_10.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient10_sd_5/wave2/implausibilities_patient10_sd_5.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/patient18_sd_10/wave2/implausibilities_patient18_sd_10.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/implausibilities_patient18_sd_5.dat", dtype=float)

print(np.max(implausibility))


In [ ]:
import postprocessing

postprocessing.compare_nroy_two_cases(folder_a="patient1_sd_10", wave_a_number=2, wave_a_name="wave2_patient1_sd_10",
                           emulators_a=["initial_sweep", "patient1_sd_10/wave1", "patient1_sd_10/wave2"],
                           biomarkers_patient_number_a=1, sd_magnitude_a=10,
                           folder_b="patient1_sd_5", wave_b_number=2, wave_b_name="wave2_patient1_sd_5",
                           emulators_b=["initial_sweep", "patient1_sd_5/wave1", "patient1_sd_5/wave2"],
                           biomarkers_patient_number_b=1, sd_magnitude_b=5)

postprocessing.compare_nroy_two_cases(folder_a="patient2_sd_10", wave_a_number=2, wave_a_name="wave2_patient2_sd_10",
                           emulators_a=["initial_sweep", "patient2_sd_10/wave1", "patient2_sd_10/wave2"],
                           biomarkers_patient_number_a=2, sd_magnitude_a=10,
                           folder_b="patient2_sd_5", wave_b_number=2, wave_b_name="wave2_patient2_sd_5",
                           emulators_b=["initial_sweep", "patient2_sd_5/wave1", "patient2_sd_5/wave2"],
                           biomarkers_patient_number_b=2, sd_magnitude_b=5)

postprocessing.compare_nroy_two_cases(folder_a="patient10_sd_10", wave_a_number=2, wave_a_name="wave2_patient10_sd_10",
                           emulators_a=["initial_sweep", "patient10_sd_10/wave1", "patient10_sd_10/wave2"],
                           biomarkers_patient_number_a=10, sd_magnitude_a=10,
                           folder_b="patient10_sd_5", wave_b_number=2, wave_b_name="wave2_patient10_sd_5",
                           emulators_b=["initial_sweep", "patient10_sd_5/wave1", "patient10_sd_5/wave2"],
                           biomarkers_patient_number_b=10, sd_magnitude_b=5)

postprocessing.compare_nroy_two_cases(folder_a="patient18_sd_10", wave_a_number=2, wave_a_name="wave2_patient18_sd_10",
                           emulators_a=["initial_sweep", "patient18_sd_10/wave1", "patient18_sd_10/wave2"],
                           biomarkers_patient_number_a=18, sd_magnitude_a=10,
                           folder_b="initial_sweep", wave_b_number=0, wave_b_name="wave0_patient18_sd_5",
                           emulators_b=["initial_sweep"],
                           biomarkers_patient_number_b=18, sd_magnitude_b=5)

# Find why the NROY of the L.U.C. is not completely included in the NROY of the H.U.C.

As a first approach, we check the overlapping of the NROY's using the emulators for the initial wave:

In [5]:
import postprocessing

postprocessing.compare_nroy_two_cases(folder_a="initial_sweep", wave_a_number=0, wave_a_name="wave0_patient1_sd_10",
                           emulators_a=["initial_sweep"],
                           biomarkers_patient_number_a=1, sd_magnitude_a=10,
                           folder_b="initial_sweep", wave_b_number=0, wave_b_name="wave0_patient1_sd_5",
                           emulators_b=["initial_sweep"],
                           biomarkers_patient_number_b=1, sd_magnitude_b=5)

89.46% of the space of initial_sweep matches that of initial_sweep
89.46% of the space of initial_sweep matches that of initial_sweep
23.22% of the NROY of initial_sweep matches that of initial_sweep
100.0% of the NROY of initial_sweep matches that of initial_sweep


[89.46, 89.46, 23.22, 100.0]

In [6]:
import postprocessing

postprocessing.compare_nroy_two_cases(folder_a="patient1_sd_10", wave_a_number=2, wave_a_name="wave2_patient1_sd_10",
                           emulators_a=["initial_sweep", "patient1_sd_10/wave1"],
                           biomarkers_patient_number_a=1, sd_magnitude_a=10,
                           folder_b="patient1_sd_5", wave_b_number=2, wave_b_name="wave2_patient1_sd_5",
                           emulators_b=["initial_sweep", "patient1_sd_5/wave1"],
                           biomarkers_patient_number_b=1, sd_magnitude_b=5)

91.13% of the space of patient1_sd_10 matches that of patient1_sd_5
91.13% of the space of patient1_sd_5 matches that of patient1_sd_10
13.59% of the NROY of patient1_sd_10 matches that of patient1_sd_5
62.24% of the NROY of patient1_sd_5 matches that of patient1_sd_10


[91.13, 91.13, 13.59, 62.24]

In [7]:
import postprocessing

postprocessing.compare_nroy_two_cases(folder_a="patient1_sd_10", wave_a_number=2, wave_a_name="wave2_patient1_sd_10",
                           emulators_a=["initial_sweep", "patient1_sd_10/wave1", "patient1_sd_10/wave2"],
                           biomarkers_patient_number_a=1, sd_magnitude_a=10,
                           folder_b="patient1_sd_5", wave_b_number=2, wave_b_name="wave2_patient1_sd_5",
                           emulators_b=["initial_sweep", "patient1_sd_5/wave1", "patient1_sd_5/wave2"],
                           biomarkers_patient_number_b=1, sd_magnitude_b=5)

90.8% of the space of patient1_sd_10 matches that of patient1_sd_5
90.8% of the space of patient1_sd_5 matches that of patient1_sd_10
7.92% of the NROY of patient1_sd_10 matches that of patient1_sd_5
55.14% of the NROY of patient1_sd_5 matches that of patient1_sd_10


[90.8, 90.8, 7.92, 55.14]

So it doesn't seem to be a bug. Looking at the VQ seems that the median variance also increases more in the L.U.C. and so points that before were implausible, now they are not implausible because we lose confindence in the ability of the emulator to predict. In fact, the variances are:

In [8]:
import postprocessing

postprocessing.print_emulator_variances(emulators_folders=["initial_sweep", "patient1_sd_10/wave1", "patient1_sd_10/wave2"])

The median variances are [22.36, 29.6, 12.15, 12.88, 7.72, 8.1, 31.1, 2.36, 13.2, 2.14, 33.73, 17.15, 12.04]


[22.361498,
 29.59712,
 12.148354,
 12.879721,
 7.7194386,
 8.103218,
 31.100185,
 2.364392,
 13.202198,
 2.135342,
 33.72928,
 17.151297,
 12.043879]

While in the L.U.C. the variances are:

In [9]:
import postprocessing

postprocessing.print_emulator_variances(emulators_folders=["initial_sweep", "patient1_sd_5/wave1", "patient1_sd_5/wave2"])

The median variances are [24.0, 34.36, 17.51, 16.01, 8.75, 9.77, 36.73, 2.95, 13.66, 2.22, 35.98, 23.02, 10.81]


[24.002808,
 34.36161,
 17.505074,
 16.005962,
 8.749542,
 9.767463,
 36.727364,
 2.9507396,
 13.658985,
 2.224835,
 35.97692,
 23.0153,
 10.807833]

In [10]:
import numpy as np
import postprocessing

HUC_var = postprocessing.print_emulator_variances(emulators_folders=["initial_sweep", "patient1_sd_10/wave1", "patient1_sd_10/wave2"])
LUC_var = postprocessing.print_emulator_variances(emulators_folders=["initial_sweep", "patient1_sd_5/wave1", "patient1_sd_5/wave2"])

delta_var = [100*(LUC_var[i]/HUC_var[i]-1) for i in range(len(HUC_var))]
print([round(d,2) for d in delta_var])

print("There has been a mean change of {}%+-{}%".format(round(np.mean(delta_var),2),round(np.std(delta_var)),2))

The median variances are [22.36, 29.6, 12.15, 12.88, 7.72, 8.1, 31.1, 2.36, 13.2, 2.14, 33.73, 17.15, 12.04]
The median variances are [24.0, 34.36, 17.51, 16.01, 8.75, 9.77, 36.73, 2.95, 13.66, 2.22, 35.98, 23.02, 10.81]
[7.34, 16.1, 44.09, 24.27, 13.34, 20.54, 18.09, 24.8, 3.46, 4.19, 6.66, 34.19, -10.26]
There has been a mean change of 15.91%+-14%


However, it seems that in the case of the ground truth points the situation changes, since in the H.U.C. they were non-implausible, while in the L.U.C. they are implausible. 

In [11]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep", "patient1_sd_10/wave1", "patient1_sd_10/wave2"],
                                       patient_number=1, sd_magnitude=10)

The implausibilities are [0.04, 0.01, 0.01, 0.02, 0.01, 0.12, 0.03, 0.07, 0.08, 0.37, 0.01, 0.23, 0.46]
The numerators are [0.41, 0.19, 0.09, 0.18, 0.03, 0.33, 0.58, 0.09, 0.33, 0.41, 0.05, 2.37, 6.0]
The denominators are [11.66, 17.92, 7.06, 11.3, 2.42, 2.86, 16.94, 1.24, 4.32, 1.13, 8.68, 10.19, 13.12]
The emulated results are [116.9, 179.15, 70.23, 112.97, 24.12, 27.96, 168.67, 12.22, 42.64, 11.57, 86.21, 85.17, 75.91]
The emulated variances are [116.9, 179.15, 70.23, 112.97, 24.12, 27.96, 168.67, 12.22, 42.64, 11.57, 86.21, 85.17, 75.91]


In [1]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep", "patient1_sd_5/wave1",
                                                                     "patient1_sd_5/wave2"],
                                       patient_number=1, sd_magnitude=5)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.18, 1.96, 2.23, 3.34, 1.06, 0.03, 1.96, 2.6, 0.59, 0.85, 1.52, 0.49, 0.52]
The numerators are [1.43, 30.83, 20.31, 36.29, 1.68, 0.05, 27.42, 2.71, 1.41, 0.57, 8.79, 3.25, 3.48]
The denominators are [7.88, 15.71, 9.1, 10.87, 1.59, 1.98, 13.97, 1.04, 2.4, 0.67, 5.78, 6.58, 6.62]
The emulated results are [117.92, 148.13, 50.01, 76.5, 22.47, 28.24, 141.83, 9.6, 44.38, 11.73, 77.37, 79.55, 78.43]
The emulated SDs are [5.31, 12.91, 8.4, 9.29, 1.03, 1.39, 11.11, 0.84, 1.08, 0.37, 3.85, 5.12, 5.21]


Therefore, in the case of subject #01 the ground truth point becomes implausible because instead of predicting a RAV of 113 mL approx, it predicts 50.01 mL. We repeat the process for subject #02:

In [ ]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep", "patient2_sd_10/wave1",
                                                                     "patient2_sd_10/wave2"],
                                       patient_number=2, sd_magnitude=10)

In [ ]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep", "patient2_sd_5/wave1",
                                                                     "patient2_sd_5/wave2"],
                                       patient_number=2, sd_magnitude=5)

In this case, the problem is with the RVOTdiam, predicting 28.45 mm instead of 35.54. As for case #10:

In [ ]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep", "patient10_sd_10/wave1",
                                                                     "patient10_sd_10/wave2"],
                                       patient_number=10, sd_magnitude=10)

In [ ]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep", "patient10_sd_5/wave1",
                                                                     "patient10_sd_5/wave2"],
                                       patient_number=10, sd_magnitude=5)

In this case, the problem is with LAV, predicting a volume of 45.3 mL, instead of 86 mL. So the problem in all cases is a loss of accuracy in the emulators. Since from the first wave there is a collapse of the NROY space and in new waves the emulators train in the neighbourhood of the NROY, the L.U.C. emulators do not have enough accuracy to predict points outside of the NROY and therefore being non-implausible. This fact provokes a domino effect on the shape of the NROY. A potential solution if using higher accuracy data acquisition techniques such as MRI would be to sample in a bigger space (not only in the neighborhood of the NROY region); reducing the number of training points in subsequent waves or run only a single wave.

# Extra edits before submission

We added a column to Table 3 repeating the analysis but using the emulators only of the initial sweep. 

NROY size as % of original space:

In [ ]:
import numpy as np

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient1_sd_10.dat", dtype=float)
print(wave0_NROY_rel)


wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient2_sd_10.dat", dtype=float)
print(wave0_NROY_rel)

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient10_sd_10.dat", dtype=float)
print(wave0_NROY_rel)

wave0_NROY_rel = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/NROY_rel_patient18_sd_10.dat", dtype=float)
print(wave0_NROY_rel)

Implausibility of simulated point:

In [ ]:
implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/implausibilities_patient1_sd_10.dat", dtype=float)

print(np.max(implausibility))

implausibility = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/implausibilities_patient2_sd_10.dat", dtype=float)

print(np.max(implausibility))

$VQ$:

In [ ]:
VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/variance_quotient_wave0_patient1_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

In [ ]:
VQ = np.genfromtxt("/media/crg17/SeagateExpansionDrive/initial_sweep/variance_quotient_wave0_patient2_sd_10.dat", dtype=float)
print(np.max(VQ))
print(np.median(VQ))

NROY+RO/NROY match:

In [ ]:
import postprocessing

perc_initial_sweep_using_2_RO = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=1, original_last_wave=0, using_patient=2,
                        using_last_wave=2)
perc_initial_sweep_using_2_NROY = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=1, original_last_wave=0, using_patient=2,
                        using_last_wave=2)
perc_initial_sweep_using_1_RO = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=True, original_patient=2, original_last_wave=0, using_patient=1,
                        using_last_wave=2)
perc_initial_sweep_using_1_NROY = postprocessing.compare_nroy_binary(n_samples=1e5, whole_space=False, original_patient=2, original_last_wave=0, using_patient=1,
                        using_last_wave=2)


print(round(perc_initial_sweep_using_2_RO,2), round(perc_initial_sweep_using_2_NROY,2), round(perc_initial_sweep_using_1_RO,2), round(perc_initial_sweep_using_1_NROY,2))

# Plot parameter space

A way to visualize the parameter space is drawing the distribution of parameter values in the NROY region. We used kernel densities estimations of histograms with a bin width of 5% the range of the values for each parameter. We first show initial space values values compared with subject #01:

In [ ]:
import postprocessing
import os
from global_variables_config import *

waves_paths = [os.path.join(PROJECT_PATH, "initial_sweep", "wave0_patient1_sd_10"),
               os.path.join(PROJECT_PATH, "patient1_sd_10", "wave1", "wave1_patient1_sd_10"),
               os.path.join(PROJECT_PATH, "patient1_sd_10", "wave2", "wave2_patient1_sd_10")
             ]
legends = ["Using emulators from the first wave of #01",
           "Using emulators from the second wave of #01",
           "Using emulators from the third wave of #01"
          ]

file_path = os.path.join(PROJECT_PATH, "figures")
file_name = "densities_waves_#01.png"

postprocessing.plot_parameter_distributions(waves_paths = waves_paths, legends = legends,
                                            file_path = file_path, file_name = file_name)

Now we compare the last wave of two very different patients and if we use the biomarkers of one with the emulators of the other:

In [ ]:
import postprocessing
import os
from global_variables_config import *

waves_paths = [os.path.join(PROJECT_PATH, "patient1_sd_10", "wave2", "wave2_patient1_sd_10"),
               os.path.join(PROJECT_PATH, "patient18_sd_10", "wave2", "wave2_patient18_sd_10"),
               os.path.join(PROJECT_PATH, "using_patient1_sd_10", "wave2", "wave2_patient18_using_patient1_sd_10")
             ]
legends = ["Trained with data from #01\n and constraining with data from #01",
           "Trained with data from #18\n and constraining with data from #18",
           "Trained with data from #01\n and constraining with data from #18",
          ]

file_path = os.path.join(PROJECT_PATH, "figures")
file_name = "densities_#01_vs_#18.png"

postprocessing.plot_parameter_distributions(waves_paths = waves_paths, legends = legends,
                                            file_path = file_path, file_name = file_name)

# Creating a 1D example

We hypothesise that training with high density of points in a region compared to low density outside increases the uncertainty in the middle. To test this we try to reproduce it using a 1D example:

In [1]:
import os
import sys
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)

In [2]:
# function to learn (normally a high-dimensional, expensive deterministic model)
from GPErks.utils.test_functions import forrester
f = lambda x: np.sin(20*x)
D = 1

In [3]:
import numpy as np

from GPErks.gp.data.dataset import Dataset
dataset_1 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=3,
    n_test_samples=0,
    design="srs",
    seed=2,
    l_bounds=[0],
    u_bounds=[0.5] 
)

In [4]:
# we add extra point at the end

dataset_2 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=4,
    n_test_samples=0,
    design="srs",
    seed=2,
    l_bounds=[0.8],
    u_bounds=[1]  
)


# we merge them

dataset = dataset_1

dataset.X_train = np.vstack([dataset_1.X_train,dataset_2.X_train])
dataset.y_train = np.append(dataset_1.y_train,dataset_2.y_train)

dataset.sample_size = dataset_1.sample_size + dataset_2.sample_size

dataset.l_bounds = [np.min([dataset_1.l_bounds[0],dataset_2.l_bounds[0]])]
dataset.u_bounds = [np.max([dataset_1.u_bounds[0],dataset_2.u_bounds[0]])]

# choose likelihood
from gpytorch.likelihoods import GaussianLikelihood
likelihood = GaussianLikelihood()

# choose mean function
from gpytorch.means import LinearMean
mean_function = LinearMean(input_size=dataset.input_size)

# choose covariance function (kernel)
from gpytorch.kernels import RBFKernel, ScaleKernel
kernel = ScaleKernel(RBFKernel(ard_num_dims=dataset.input_size))

# choose metrics
from torchmetrics import MeanSquaredError, R2Score
metrics = [MeanSquaredError(), R2Score()]

# define experiment
from GPErks.gp.experiment import GPExperiment
experiment = GPExperiment(
    dataset,
    likelihood,
    mean_function,
    kernel,
    n_restarts=3,
    metrics=metrics,
    seed=2,  # reproducible training
    learn_noise=True  # y = f(x) + e, e ~ N(0, sigma^2I)
)

# train model
from GPErks.train.emulator import GPEmulator
import torch
emulator = GPEmulator(experiment, device="cpu")
optimizer = torch.optim.Adam(experiment.model.parameters(), lr=0.1)
emulator.train(optimizer)

import numpy as np

x_train = dataset.X_train
y_train = dataset.y_train

xx = np.linspace(dataset.l_bounds[0], dataset.u_bounds[0], 1000)

yy_mean, yy_std = emulator.predict(xx)
yy_true = f(xx)

import matplotlib.pyplot as plt
height = 9.36111
width = 5.91667
fig, axis = plt.subplots(1, 1, figsize=(4*width/3, height/2))

axis.plot(xx, yy_true, c="C0", ls="--", label="true function")

CI = 2
axis.plot(xx, yy_mean, c="C0", label="predicted mean")
axis.fill_between(
    xx, yy_mean - CI * yy_std, yy_mean + CI * yy_std, color="C0", alpha=0.15, label="~95% CI"
)
axis.scatter(x_train, y_train, fc="C0", ec="C0", label="training data")
axis.set_ylim([-2,4])
axis.legend(loc="best")
fig.tight_layout()
plt.show()

print(np.max(yy_std[200:900]))

In [5]:
# we add extra point at the end

dataset_2 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=10,
    n_test_samples=0,
    design="srs",
    seed=2,
    l_bounds=[0.8],
    u_bounds=[1]  
)


# we merge them

dataset = dataset_1

dataset.X_train = np.vstack([dataset_1.X_train,dataset_2.X_train])
dataset.y_train = np.append(dataset_1.y_train,dataset_2.y_train)

dataset.sample_size = dataset_1.sample_size + dataset_2.sample_size

dataset.l_bounds = [np.min([dataset_1.l_bounds[0],dataset_2.l_bounds[0]])]
dataset.u_bounds = [np.max([dataset_1.u_bounds[0],dataset_2.u_bounds[0]])]

# choose likelihood
from gpytorch.likelihoods import GaussianLikelihood
likelihood = GaussianLikelihood()

# choose mean function
from gpytorch.means import LinearMean
mean_function = LinearMean(input_size=dataset.input_size)

# choose covariance function (kernel)
from gpytorch.kernels import RBFKernel, ScaleKernel
kernel = ScaleKernel(RBFKernel(ard_num_dims=dataset.input_size))

# choose metrics
from torchmetrics import MeanSquaredError, R2Score
metrics = [MeanSquaredError(), R2Score()]

# define experiment
from GPErks.gp.experiment import GPExperiment
experiment = GPExperiment(
    dataset,
    likelihood,
    mean_function,
    kernel,
    n_restarts=3,
    metrics=metrics,
    seed=2,  # reproducible training
    learn_noise=True  # y = f(x) + e, e ~ N(0, sigma^2I)
)

# train model
from GPErks.train.emulator import GPEmulator
import torch
emulator = GPEmulator(experiment, device="cpu")
optimizer = torch.optim.Adam(experiment.model.parameters(), lr=0.1)
emulator.train(optimizer)

import numpy as np

x_train = dataset.X_train
y_train = dataset.y_train

xx = np.linspace(dataset.l_bounds[0], dataset.u_bounds[0], 1000)

yy_mean, yy_std = emulator.predict(xx)
yy_true = f(xx)

import matplotlib.pyplot as plt
height = 9.36111
width = 5.91667
fig, axis = plt.subplots(1, 1, figsize=(4*width/3, height/2))

axis.plot(xx, yy_true, c="C0", ls="--", label="true function")

CI = 2
axis.plot(xx, yy_mean, c="C0", label="predicted mean")
axis.fill_between(
    xx, yy_mean - CI * yy_std, yy_mean + CI * yy_std, color="C0", alpha=0.15, label="~95% CI"
)
axis.scatter(x_train, y_train, fc="C0", ec="C0", label="training data")
axis.set_ylim([-2,4])
axis.legend(loc="best")
fig.tight_layout()
plt.show()

print(np.max(yy_std[200:900]))

# Trying to make the uncertainty not capture the ground truth

In [24]:
import os
import sys
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
import numpy as np
    
# function to learn (normally a high-dimensional, expensive deterministic model)
from GPErks.utils.test_functions import forrester
# a_sin = 120.
# f = lambda x: np.sin(a_sin*x)
alpha_fun = 0.4
omega_fun = 20
beta_fun = 0.4
c_fun = -0.6
f = lambda x: alpha_fun*np.cos((omega_fun*x)/beta_fun)-c_fun*(x/beta_fun)**2
D = 1


from GPErks.gp.data.dataset import Dataset

dataset_1 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=2,
    n_test_samples=1,
    design="srs",
    seed=2,
    l_bounds=[-1],
    u_bounds=[-0.75] 
)

# we add extra point at the end

dataset_2 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=2,
    n_test_samples=0,
    design="srs",
    seed=2,
    l_bounds=[0.93],
    u_bounds=[1]  
)

dataset_3 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=2,
    n_test_samples=0,
    design="srs",
    seed=1,
#     l_bounds=[2*2*np.pi/a_sin-0.01-0.25*2*np.pi/a_sin-0.055],
#     u_bounds=[2*2*np.pi/a_sin+0.01-0.25*2*np.pi/a_sin-0.055]  
    l_bounds = [-0.35],
    u_bounds = [-0.3]
)


# we merge them

dataset = dataset_1

dataset.X_train = np.vstack([dataset_1.X_train,dataset_2.X_train,dataset_3.X_train])
dataset.y_train = np.concatenate((dataset_1.y_train,dataset_2.y_train,dataset_3.y_train))

dataset.sample_size = dataset_1.sample_size + dataset_2.sample_size + dataset_3.sample_size

dataset.l_bounds = [np.min([dataset_1.l_bounds[0],dataset_2.l_bounds[0],dataset_3.l_bounds[0]])]
dataset.u_bounds = [np.max([dataset_1.u_bounds[0],dataset_2.u_bounds[0],dataset_3.u_bounds[0]])]

# dataset.X_train = np.vstack([dataset_1.X_train,dataset_2.X_train])
# dataset.y_train = np.append(dataset_1.y_train,[dataset_2.y_train])

# dataset.sample_size = dataset_1.sample_size + dataset_2.sample_size 

# dataset.l_bounds = [np.min([dataset_1.l_bounds[0],dataset_2.l_bounds[0]])]
# dataset.u_bounds = [np.max([dataset_1.u_bounds[0],dataset_2.u_bounds[0]])]

# choose likelihood
from gpytorch.likelihoods import GaussianLikelihood
likelihood = GaussianLikelihood()

# choose mean function
from gpytorch.means import LinearMean
mean_function = LinearMean(input_size=dataset.input_size)

# choose covariance function (kernel)
from gpytorch.kernels import RBFKernel, ScaleKernel
kernel = ScaleKernel(RBFKernel(ard_num_dims=dataset.input_size))

# choose metrics
from torchmetrics import MeanSquaredError, R2Score
metrics = [MeanSquaredError(), R2Score()]

# define experiment
from GPErks.gp.experiment import GPExperiment
experiment = GPExperiment(
    dataset,
    likelihood,
    mean_function,
    kernel,
    n_restarts=3,
    metrics=metrics,
    seed=2,  # reproducible training
    learn_noise=True  # y = f(x) + e, e ~ N(0, sigma^2I)
)

# train model
from GPErks.train.emulator import GPEmulator
import torch
emulator = GPEmulator(experiment, device="cpu")
optimizer = torch.optim.Adam(experiment.model.parameters(), lr=0.1)
emulator.train(optimizer)

import numpy as np

x_train = dataset.X_train
y_train = dataset.y_train

xx = np.linspace(dataset.l_bounds[0], dataset.u_bounds[0], 1000)

yy_mean, yy_std = emulator.predict(xx)
yy_true = f(xx)

import matplotlib.pyplot as plt
height = 9.36111
width = 5.91667
fig, axis = plt.subplots(1, 1, figsize=(4*width/3, height/2))

axis.plot(xx, yy_true, c="C0", ls="--", label="true function")

CI = 3
axis.plot(xx, yy_mean, c="C0", label="predicted mean")
axis.fill_between(
    xx, yy_mean - CI * yy_std, yy_mean + CI * yy_std, color="C0", alpha=0.15, label="~95% CI"
)
axis.scatter(x_train, y_train, fc="C0", ec="C0", label="training data")
axis.set_ylim([-1,4])
axis.legend(loc="best")
fig.tight_layout()
plt.show()

print(np.max(yy_std[200:900]))

In [ ]:
import os
import sys
module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
import numpy as np
    
# function to learn (normally a high-dimensional, expensive deterministic model)
from GPErks.utils.test_functions import forrester
# a_sin = 120.
# f = lambda x: np.sin(a_sin*x)
alpha_fun = 0.4
omega_fun = 20
beta_fun = 0.4
c_fun = -0.6
f = lambda x: alpha_fun*np.cos((omega_fun*x)/beta_fun)-c_fun*(x/beta_fun)**2
D = 1


from GPErks.gp.data.dataset import Dataset

dataset_1 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=2,
    n_test_samples=1,
    design="srs",
    seed=2,
    l_bounds=[-1],
    u_bounds=[-0.75] 
)

# we add extra point at the end

dataset_2 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=20,
    n_test_samples=0,
    design="srs",
    seed=2,
    l_bounds=[0.99],
    u_bounds=[1]  
)

dataset_3 = Dataset.build_from_function(
    f,
    D,
    n_train_samples=2,
    n_test_samples=0,
    design="srs",
    seed=1,
#     l_bounds=[2*2*np.pi/a_sin-0.01-0.25*2*np.pi/a_sin-0.055],
#     u_bounds=[2*2*np.pi/a_sin+0.01-0.25*2*np.pi/a_sin-0.055]  
    l_bounds = [-0.35],
    u_bounds = [-0.3]
)


# we merge them

dataset = dataset_1

dataset.X_train = np.vstack([dataset_1.X_train,dataset_2.X_train,dataset_3.X_train])
dataset.y_train = np.concatenate((dataset_1.y_train,dataset_2.y_train,dataset_3.y_train))

dataset.sample_size = dataset_1.sample_size + dataset_2.sample_size + dataset_3.sample_size

dataset.l_bounds = [np.min([dataset_1.l_bounds[0],dataset_2.l_bounds[0],dataset_3.l_bounds[0]])]
dataset.u_bounds = [np.max([dataset_1.u_bounds[0],dataset_2.u_bounds[0],dataset_3.u_bounds[0]])]

# dataset.X_train = np.vstack([dataset_1.X_train,dataset_2.X_train])
# dataset.y_train = np.append(dataset_1.y_train,[dataset_2.y_train])

# dataset.sample_size = dataset_1.sample_size + dataset_2.sample_size 

# dataset.l_bounds = [np.min([dataset_1.l_bounds[0],dataset_2.l_bounds[0]])]
# dataset.u_bounds = [np.max([dataset_1.u_bounds[0],dataset_2.u_bounds[0]])]

# choose likelihood
from gpytorch.likelihoods import GaussianLikelihood
likelihood = GaussianLikelihood()

# choose mean function
from gpytorch.means import LinearMean
mean_function = LinearMean(input_size=dataset.input_size)

# choose covariance function (kernel)
from gpytorch.kernels import RBFKernel, ScaleKernel
kernel = ScaleKernel(RBFKernel(ard_num_dims=dataset.input_size))

# choose metrics
from torchmetrics import MeanSquaredError, R2Score
metrics = [MeanSquaredError(), R2Score()]

# define experiment
from GPErks.gp.experiment import GPExperiment
experiment = GPExperiment(
    dataset,
    likelihood,
    mean_function,
    kernel,
    n_restarts=3,
    metrics=metrics,
    seed=2,  # reproducible training
    learn_noise=True  # y = f(x) + e, e ~ N(0, sigma^2I)
)

# train model
from GPErks.train.emulator import GPEmulator
import torch
emulator = GPEmulator(experiment, device="cpu")
optimizer = torch.optim.Adam(experiment.model.parameters(), lr=0.1)
emulator.train(optimizer)

import numpy as np

x_train = dataset.X_train
y_train = dataset.y_train

xx = np.linspace(dataset.l_bounds[0], dataset.u_bounds[0], 1000)

yy_mean, yy_std = emulator.predict(xx)
yy_true = f(xx)

import matplotlib.pyplot as plt
height = 9.36111
width = 5.91667
fig, axis = plt.subplots(1, 1, figsize=(4*width/3, height/2))

axis.plot(xx, yy_true, c="C0", ls="--", label="true function")

CI = 2
axis.plot(xx, yy_mean, c="C0", label="predicted mean")
axis.fill_between(
    xx, yy_mean - CI * yy_std, yy_mean + CI * yy_std, color="C0", alpha=0.15, label="~95% CI"
)
axis.scatter(x_train, y_train, fc="C0", ec="C0", label="training data")
axis.set_ylim([-1,5])
axis.legend(loc="best")
fig.tight_layout()
plt.show()

print(np.max(yy_std[200:900]))

# Extra efforts to explain why LUC is not in HUC and why the ground truth points are not in the NROY

To try to find a more satisfactory explanation to the problems with the LUC, we added more cases to the global emulator. We merged manually the input points of the initial emulator with those of patients 2, 10 and 18 for waves 1 and 2. We did the same with all the biomarkers for a total of 1120 points, instead of 280. We now train the emulators:

In [ ]:
import emulators

emulators_vector = emulators.train(folders=["initial_sweep_all_patients_but_1"])

And now we do the BHM:

In [7]:
import history_matching

wave = history_matching.compute_nroy_region(emulators_vector=emulators_vector, implausibility_threshold=3.2,
                                                    literature_data=False, input_folder="initial_sweep_all_patients_but_1",
                                                    patient_number=1, sd_magnitude=5,
                                                    first_time=True)

Computing implausibility of NROY...
Finished
Results saved in /media/crg17/SeagateExpansionDrive/initial_sweep_all_patients_but_1


In [ ]:
history_matching.plot_nroy(input_folder="initial_sweep_all_patients_but_1", wave=wave, literature_data=False,
                                   patient_number=1, sd_magnitude=5,
                                   title = "Initial wave for #1 with SD=5% using data from all previous patients")

In [10]:
history_matching.generate_new_training_pts(wave=wave, num_pts=140, output_folder="patient1_sd_5_all_patients/wave1",
                                                   input_folder="initial_sweep_all_patients_but_1", wave_name="wave0_patient1_sd_5")

We check how is it going:

In [15]:
import postprocessing

postprocessing.compare_nroy_two_cases(folder_a="initial_sweep_all_patients_but_1", wave_a_number=0, wave_a_name="wave0_patient1_sd_5",
                           emulators_a=["initial_sweep_all_patients_but_1"],
                           biomarkers_patient_number_a=1, sd_magnitude_a=5,
                           folder_b="initial_sweep", wave_b_number=0, wave_b_name="wave0_patient1_sd_10",
                           emulators_b=["initial_sweep"],
                           biomarkers_patient_number_b=1, sd_magnitude_b=10)

89.58% of the space of initial_sweep_all_patients_but_1 matches that of initial_sweep
89.58% of the space of initial_sweep matches that of initial_sweep_all_patients_but_1
87.8% of the NROY of initial_sweep_all_patients_but_1 matches that of initial_sweep
24.53% of the NROY of initial_sweep matches that of initial_sweep_all_patients_but_1


[89.58, 89.58, 87.8, 24.53]

In [17]:
postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep_all_patients_but_1"],
                                       patient_number=1, sd_magnitude=5)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.08, 0.03, 0.18, 0.04, 0.02, 0.08, 0.13, 0.17, 0.26, 0.65, 0.04, 1.51, 0.08]
The numerators are [0.5, 0.23, 0.66, 0.25, 0.02, 0.12, 1.13, 0.11, 0.58, 0.38, 0.16, 10.21, 0.65]
The denominators are [5.86, 8.99, 3.59, 5.69, 1.22, 1.49, 8.5, 0.64, 2.21, 0.59, 4.42, 6.75, 8.56]
The emulated results are [116.99, 178.73, 69.66, 112.54, 24.17, 28.17, 168.12, 12.2, 42.39, 11.54, 86.0, 93.01, 82.56]
The emulated SDs are [0.65, 0.9, 0.73, 0.77, 0.17, 0.47, 0.78, 0.19, 0.5, 0.19, 0.99, 5.33, 7.51]


The problem is that we cannot compare them directly, because the ground truth of the naive approach with HUC, has a different training set. We would need to run with SD=10% but with this new training set. Since the implausibilities of the ground truth points are at least under the threshold for now we continue with SD=5%.

In [20]:
import generate_meshes
import preprocess_mesh
import ep_simulations
import biomarkers
import emulators
import history_matching
import os

from global_variables_config import *

print("Running wave 1...")

generate_meshes.sample_atlas(subfolder="patient1_sd_5_all_patients/wave1",
                             csv_filename="input_anatomy_training.csv")
preprocess_mesh.biv_setup(subfolder="patient1_sd_5_all_patients/wave1",
                          anatomy_csv_file="input_anatomy_training.csv",
                          ep_dat_file="input_ep_training.dat")
ep_simulations.run(subfolder="patient1_sd_5_all_patients/wave1",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")
biomarkers.extract(subfolder="patient1_sd_5_all_patients/wave1",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")

emulators_vector = emulators.train(folders=["initial_sweep_all_patients_but_1","patient1_sd_5_all_patients/wave1"])
history_matching.save_patient_implausibility(emulators_vector=emulators_vector, 
                                             input_folder="patient1_sd_5_all_patients/wave1",
                                             patient_number=1, sd_magnitude=5)
wave = history_matching.compute_nroy_region(emulators_vector=emulators_vector, 
                                            implausibility_threshold=3.2,
                                            literature_data=False, 
                                            input_folder="patient1_sd_5_all_patients/wave1",
                                            patient_number=1, 
                                            sd_magnitude=5,
                                            previous_wave_name=os.path.join(PROJECT_PATH,"initial_sweep_all_patients_but_1","wave0_patient1_sd_5"))
history_matching.plot_nroy(input_folder="patient1_sd_5_all_patients/wave1", 
                           wave=wave, 
                           literature_data=False,
                           patient_number=1, 
                           sd_magnitude=5, 
                           title = "Second wave for #1 with SD=5% using data from all previous patients")
history_matching.generate_new_training_pts(wave=wave, num_pts=140, 
                                           output_folder="patient1_sd_5_all_patients/wave2",
                                           input_folder="patient1_sd_5_all_patients/wave1",
                                           wave_name="wave1_patient1_sd_5")

We check how is it going with the ground truth:

In [25]:
postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep_all_patients_but_1","patient1_sd_5_all_patients/wave1"],
                                       patient_number=1, sd_magnitude=5)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.07, 2.15, 2.66, 2.99, 1.14, 0.93, 1.6, 2.0, 0.43, 0.62, 1.4, 0.22, 0.03]
The numerators are [0.49, 37.6, 23.81, 40.25, 1.93, 1.71, 19.92, 1.94, 1.04, 0.41, 8.92, 1.84, 0.26]
The denominators are [6.64, 17.49, 8.96, 13.45, 1.68, 1.85, 12.47, 0.97, 2.42, 0.66, 6.38, 8.54, 10.3]
The emulated results are [116.0, 141.36, 46.51, 72.54, 22.22, 26.58, 149.33, 10.37, 44.01, 11.57, 77.24, 80.96, 82.17]
The emulated SDs are [3.18, 15.02, 8.24, 12.21, 1.17, 1.19, 9.16, 0.75, 1.11, 0.35, 4.7, 7.46, 9.45]


In this step is used to fail with implausibility >3.2 but now is 2.99 so all good so far. We run the last wave:

In [26]:
import generate_meshes
import preprocess_mesh
import ep_simulations
import biomarkers
import emulators
import history_matching
import os

from global_variables_config import *

print("Running wave 2...")

generate_meshes.sample_atlas(subfolder="patient1_sd_5_all_patients/wave2",
                             csv_filename="input_anatomy_training.csv")
preprocess_mesh.biv_setup(subfolder="patient1_sd_5_all_patients/wave2",
                          anatomy_csv_file="input_anatomy_training.csv",
                          ep_dat_file="input_ep_training.dat")
ep_simulations.run(subfolder="patient1_sd_5_all_patients/wave2",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")
biomarkers.extract(subfolder="patient1_sd_5_all_patients/wave2",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")

emulators_vector = emulators.train(folders=["initial_sweep_all_patients_but_1","patient1_sd_5_all_patients/wave1","patient1_sd_5_all_patients/wave2"])
history_matching.save_patient_implausibility(emulators_vector=emulators_vector, 
                                             input_folder="patient1_sd_5_all_patients/wave2",
                                             patient_number=1, sd_magnitude=5)
wave = history_matching.compute_nroy_region(emulators_vector=emulators_vector, 
                                            implausibility_threshold=3.,
                                            literature_data=False, 
                                            input_folder="patient1_sd_5_all_patients/wave2",
                                            patient_number=1, 
                                            sd_magnitude=5,
                                            previous_wave_name=os.path.join(PROJECT_PATH,"patient1_sd_5_all_patients","wave1","wave1_patient1_sd_5"))
history_matching.plot_nroy(input_folder="patient1_sd_5_all_patients/wave2", 
                           wave=wave, 
                           literature_data=False,
                           patient_number=1, 
                           sd_magnitude=5, 
                           title = "Third wave for #1 with SD=5% using data from all previous patients")
history_matching.generate_new_training_pts(wave=wave, num_pts=140, 
                                           output_folder="patient1_sd_5_all_patients/wave3",
                                           input_folder="patient1_sd_5_all_patients/wave2",
                                           wave_name="wave2_patient1_sd_5")

Final check with the implausibility of the ground truth points

In [27]:
postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep_all_patients_but_1","patient1_sd_5_all_patients/wave1","patient1_sd_5_all_patients/wave2"],
                                       patient_number=1, sd_magnitude=5)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.09, 2.24, 2.36, 3.35, 1.19, 0.92, 1.71, 2.54, 0.74, 1.15, 1.91, 0.19, 0.0]
The numerators are [0.68, 34.23, 23.33, 41.27, 2.01, 1.87, 22.11, 2.57, 1.78, 0.82, 11.7, 1.53, 0.04]
The denominators are [7.36, 15.3, 9.9, 12.32, 1.69, 2.02, 12.94, 1.01, 2.39, 0.71, 6.14, 8.06, 9.92]
The emulated results are [117.17, 144.73, 46.99, 71.52, 22.14, 26.42, 147.14, 9.74, 44.75, 11.98, 74.46, 81.27, 81.95]
The emulated SDs are [4.5, 12.41, 9.25, 10.96, 1.18, 1.44, 9.79, 0.81, 1.05, 0.44, 4.37, 6.91, 9.04]


In [4]:
postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep_all_patients_but_1","patient1_sd_5_all_patients/wave1"],
                                       patient_number=1, sd_magnitude=5)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.07, 2.12, 2.66, 3.0, 1.15, 0.92, 1.6, 1.97, 0.43, 0.62, 1.4, 0.21, 0.03]
The numerators are [0.49, 37.6, 23.81, 40.25, 1.93, 1.71, 19.92, 1.94, 1.04, 0.41, 8.92, 1.84, 0.26]
The denominators are [6.64, 17.72, 8.94, 13.44, 1.67, 1.86, 12.46, 0.98, 2.41, 0.67, 6.39, 8.8, 10.14]
The emulated results are [116.0, 141.36, 46.51, 72.54, 22.22, 26.58, 149.33, 10.37, 44.01, 11.57, 77.24, 80.96, 82.17]
The emulated SDs are [3.19, 15.3, 8.22, 12.2, 1.16, 1.21, 9.15, 0.77, 1.1, 0.36, 4.72, 7.76, 9.27]


In [5]:
postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep","patient1_sd_5/wave1"],
                                       patient_number=1, sd_magnitude=5)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.02, 2.87, 3.27, 3.63, 1.53, 0.55, 2.14, 2.51, 0.5, 0.98, 1.75, 0.75, 1.06]
The numerators are [0.11, 41.66, 24.95, 38.65, 2.31, 1.09, 27.85, 2.49, 1.17, 0.74, 9.72, 4.81, 7.66]
The denominators are [7.13, 14.53, 7.64, 10.65, 1.51, 1.98, 13.01, 0.99, 2.36, 0.76, 5.54, 6.4, 7.2]
The emulated results are [116.6, 137.3, 45.37, 74.14, 21.84, 27.2, 141.4, 9.82, 44.14, 11.9, 76.44, 77.99, 74.25]
The emulated SDs are [4.12, 11.44, 6.78, 9.03, 0.91, 1.39, 9.88, 0.78, 0.97, 0.51, 3.48, 4.88, 5.93]


We try now rerunning it but with SD=10%. Wave 0:

In [ ]:
import emulators

emulators_vector = emulators.train(folders=["initial_sweep_all_patients_but_1"])

import history_matching

wave = history_matching.compute_nroy_region(emulators_vector=emulators_vector, implausibility_threshold=3.2,
                                                    literature_data=False, input_folder="initial_sweep_all_patients_but_1",
                                                    patient_number=1, sd_magnitude=10,
                                                    first_time=True)

history_matching.plot_nroy(input_folder="initial_sweep_all_patients_but_1", wave=wave, literature_data=False,
                                   patient_number=1, sd_magnitude=10,
                                   title = "Initial wave for #1 with SD=10% using data from all previous patients")

history_matching.generate_new_training_pts(wave=wave, num_pts=140, output_folder="patient1_sd_10_all_patients/wave1",
                                                   input_folder="initial_sweep_all_patients_but_1", wave_name="wave0_patient1_sd_10")

Now wave 1:

In [2]:
import generate_meshes
import preprocess_mesh
import ep_simulations
import biomarkers
import emulators
import history_matching
import os

from global_variables_config import *

print("Running wave 1...")

generate_meshes.sample_atlas(subfolder="patient1_sd_10_all_patients/wave1",
                             csv_filename="input_anatomy_training.csv")
preprocess_mesh.biv_setup(subfolder="patient1_sd_10_all_patients/wave1",
                          anatomy_csv_file="input_anatomy_training.csv",
                          ep_dat_file="input_ep_training.dat")
ep_simulations.run(subfolder="patient1_sd_10_all_patients/wave1",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")
biomarkers.extract(subfolder="patient1_sd_10_all_patients/wave1",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")

emulators_vector = emulators.train(folders=["initial_sweep_all_patients_but_1","patient1_sd_10_all_patients/wave1"])
history_matching.save_patient_implausibility(emulators_vector=emulators_vector, 
                                             input_folder="patient1_sd_10_all_patients/wave1",
                                             patient_number=1, sd_magnitude=10)
wave = history_matching.compute_nroy_region(emulators_vector=emulators_vector, 
                                            implausibility_threshold=3.2,
                                            literature_data=False, 
                                            input_folder="patient1_sd_10_all_patients/wave1",
                                            patient_number=1, 
                                            sd_magnitude=10,
                                            previous_wave_name=os.path.join(PROJECT_PATH,"initial_sweep_all_patients_but_1","wave0_patient1_sd_10"))
history_matching.plot_nroy(input_folder="patient1_sd_10_all_patients/wave1", 
                           wave=wave, 
                           literature_data=False,
                           patient_number=1, 
                           sd_magnitude=10, 
                           title = "Second wave for #1 with SD=10% using data from all previous patients")
history_matching.generate_new_training_pts(wave=wave, num_pts=140, 
                                           output_folder="patient1_sd_10_all_patients/wave2",
                                           input_folder="patient1_sd_10_all_patients/wave1",
                                           wave_name="wave1_patient1_sd_10")

100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 196/196 [00:37<00:00,  5.25it/s]


And the last wave:

In [ ]:
import generate_meshes
import preprocess_mesh
import ep_simulations
import biomarkers
import emulators
import history_matching
import os

from global_variables_config import *

print("Running wave 2...")

generate_meshes.sample_atlas(subfolder="patient1_sd_10_all_patients/wave2",
                             csv_filename="input_anatomy_training.csv")
preprocess_mesh.biv_setup(subfolder="patient1_sd_10_all_patients/wave2",
                          anatomy_csv_file="input_anatomy_training.csv",
                          ep_dat_file="input_ep_training.dat")
ep_simulations.run(subfolder="patient1_sd_10_all_patients/wave2",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")
biomarkers.extract(subfolder="patient1_sd_10_all_patients/wave2",
                   anatomy_csv_file="input_anatomy_training.csv",
                   ep_dat_file="input_ep_training.dat")

emulators_vector = emulators.train(folders=["initial_sweep_all_patients_but_1","patient1_sd_10_all_patients/wave1","patient1_sd_10_all_patients/wave2"])
history_matching.save_patient_implausibility(emulators_vector=emulators_vector, 
                                             input_folder="patient1_sd_10_all_patients/wave2",
                                             patient_number=1, sd_magnitude=10)
wave = history_matching.compute_nroy_region(emulators_vector=emulators_vector, 
                                            implausibility_threshold=3.,
                                            literature_data=False, 
                                            input_folder="patient1_sd_10_all_patients/wave2",
                                            patient_number=1, 
                                            sd_magnitude=10,
                                            previous_wave_name=os.path.join(PROJECT_PATH,"patient1_sd_10_all_patients","wave1","wave1_patient1_sd_10"))
history_matching.plot_nroy(input_folder="patient1_sd_10_all_patients/wave2", 
                           wave=wave, 
                           literature_data=False,
                           patient_number=1, 
                           sd_magnitude=10, 
                           title = "Third wave for #1 with SD=10% using data from all previous patients")
history_matching.generate_new_training_pts(wave=wave, num_pts=140, 
                                           output_folder="patient1_sd_10_all_patients/wave3",
                                           input_folder="patient1_sd_10_all_patients/wave2",
                                           wave_name="wave2_patient1_sd_10")

In [2]:
import postprocessing

postprocessing.print_patient_implausibility_terms(emulators_folders=["initial_sweep_all_patients_but_1","patient1_sd_10_all_patients/wave1","patient1_sd_10_all_patients/wave2"],
                                       patient_number=1, sd_magnitude=10)

The biomarkers are: 
['LVV' 'RVV' 'LAV' 'RAV' 'LVOTdiam' 'RVOTdiam' 'LVmass' 'LVWT' 'LVEDD'
 'SeptumWT' 'RVlongdiam' 'TAT' 'TATLVendo']
The implausibilities are [0.05, 0.39, 0.41, 0.89, 0.26, 0.21, 0.33, 0.22, 0.1, 0.61, 0.33, 0.27, 0.14]
The numerators are [0.69, 9.79, 4.95, 13.8, 0.73, 0.68, 7.11, 0.34, 0.43, 0.8, 3.23, 3.03, 1.83]
The denominators are [14.57, 25.09, 12.01, 15.49, 2.81, 3.25, 21.64, 1.54, 4.47, 1.3, 9.86, 11.39, 13.55]
The emulated results are [115.8, 169.17, 65.37, 98.99, 23.42, 27.61, 162.14, 11.97, 42.54, 11.96, 82.93, 85.83, 83.74]
The emulated SDs are [8.75, 17.59, 9.74, 10.61, 1.43, 1.59, 13.48, 0.93, 1.23, 0.67, 4.79, 7.82, 10.79]


In [3]:
import postprocessing

postprocessing.compare_nroy_two_cases(folder_a="patient1_sd_5_all_patients", wave_a_number=2, wave_a_name="wave2_patient1_sd_5",
                           emulators_a=["initial_sweep_all_patients_but_1","patient1_sd_5_all_patients/wave1","patient1_sd_5_all_patients/wave2"],
                           biomarkers_patient_number_a=1, sd_magnitude_a=5,
                           folder_b="patient1_sd_10_all_patients", wave_b_number=2, wave_b_name="wave2_patient1_sd_10",
                           emulators_b=["initial_sweep_all_patients_but_1","patient1_sd_10_all_patients/wave1","patient1_sd_10_all_patients/wave2"],
                           biomarkers_patient_number_b=1, sd_magnitude_b=10)

75.97% of the space of patient1_sd_5_all_patients matches that of patient1_sd_10_all_patients
75.97% of the space of patient1_sd_10_all_patients matches that of patient1_sd_5_all_patients
98.15% of the NROY of patient1_sd_5_all_patients matches that of patient1_sd_10_all_patients
11.78% of the NROY of patient1_sd_10_all_patients matches that of patient1_sd_5_all_patients


[75.97, 75.97, 98.15, 11.78]